cpp/src/io/fst/dispatch_dfa.cuh

```diff
@@ -0,0 +1,456 @@
+
+    return error;
+  }
+
+  /**
+   * @brief Computes the state-transition vectors
+   */
+  template <typename ActivePolicyT,
+            typename TileStateT,
+            typename FstScanTileStateT,
+            typename StateVectorT>
+  CUB_RUNTIME_FUNCTION __forceinline__ cudaError_t
+  ComputeStateTransitionVector(uint32_t sm_count,
+                               TileStateT tile_state,
+                               FstScanTileStateT fst_tile_state,
+                               StateVectorT* d_thread_state_transition)
+  {
+    StateIndexT seed_state = 0;
+
+    return InvokeDFASimulationKernel<ActivePolicyT>(
+      SimulateDFAKernel<true,
+                        SINGLE_PASS_STV,
+                        DfaT,
+                        TileStateT,
+                        typename ActivePolicyT::AgentDFAPolicy,
+                        SymbolItT,
+                        OffsetT,
+                        StateVectorT,
+                        FstScanTileStateT,
+                        TransducedOutItT,
+                        TransducedIndexOutItT,
+                        TransducedCountOutItT>,
+      sm_count,
+      seed_state,
+      d_thread_state_transition,
+      tile_state,
+      fst_tile_state);
+  }
+
+  /**
+   * @brief Performs the actual DFA simulation.
+   */
+  template <typename ActivePolicyT,
+            typename TileStateT,
+            typename FstScanTileStateT,
+            typename StateVectorT>
+  CUB_RUNTIME_FUNCTION __forceinline__ cudaError_t
+  SimulateDFA(uint32_t sm_count,
+              TileStateT tile_state,
+              FstScanTileStateT fst_tile_state,
+              StateIndexT seed_state,
+              StateVectorT* d_thread_state_transition)
+  {
+    return InvokeDFASimulationKernel<ActivePolicyT>(
+      SimulateDFAKernel<false,
+                        SINGLE_PASS_STV,
+                        DfaT,
+                        TileStateT,
+                        typename ActivePolicyT::AgentDFAPolicy,
+                        SymbolItT,
+                        OffsetT,
+                        StateVectorT,
+                        FstScanTileStateT,
+                        TransducedOutItT,
+                        TransducedIndexOutItT,
+                        TransducedCountOutItT>,
+      sm_count,
+      seed_state,
+      d_thread_state_transition,
+      tile_state,
+      fst_tile_state);
+  }
+
+  //------------------------------------------------------------------------------
+  // POLICY INVOKATION
+  //------------------------------------------------------------------------------
+  template <typename ActivePolicyT>
+  CUB_RUNTIME_FUNCTION __forceinline__ cudaError_t Invoke()
+  {
+    cudaError_t error = cudaSuccess;
+
+    // Get SM count
+    int device_ordinal;
+    int sm_count;
+
+    // Get current device
+    error = cudaGetDevice(&device_ordinal);
+    if (error)
+
+      error = cudaDeviceGetAttribute(&sm_count, cudaDevAttrMultiProcessorCount, device_ordinal);
```

Comment:
Suspicious formatting; my guess is that the return statement is missing from the branch (that line 320 should not be in the branch).
Should we initialize `device_ordina`l and `sm_count` to invalid values to prevent errors like this?

Reply:
Oh that was a good catch. Thanks 👍